// connect-history-api-fallback-exclusions/lib/index.js
const url = require('url');

exports = module.exports = function historyApiFallback(options) {
  options = options || {};
  const logger = getLogger(options);

  return function(req, res, next) {
    const headers = req.headers;
    if (req.method !== 'GET') {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the method is not GET.'
      );
      return next();
    } else if (!headers || typeof headers.accept !== 'string') {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the client did not send an HTTP accept header.'
      );
      return next();
    } else if (headers.accept.indexOf('application/json') === 0) {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the client prefers JSON.'
      );
      return next();
    } else if (!acceptsHtml(headers.accept, options)) {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the client does not accept HTML.'
      );
      return next();
    }

    const parsedUrl = url.parse(req.url);
    let rewriteTarget;
    options.rewrites = options.rewrites || [];
    if (isArray(options.rewrites)) {
      for (let i = 0; i < options.rewrites.length; i++) {
        let rewrite = options.rewrites[i];
        let match = parsedUrl.pathname.match(rewrite.from);
        if (match !== null) {
          rewriteTarget = evaluateRewriteRule(parsedUrl, match, rewrite.to, req);
          logger('Rewriting', req.method, req.url, 'to', rewriteTarget);
          req.url = rewriteTarget;
          return next();
        }
      }
    }

    options.exclusions = options.exclusions || [];
    if (isArray(options.exclusions)) {
      for (let i = 0; i < options.exclusions.length; i++) {
        let exclusion = options.exclusions[i];
        let match = parsedUrl.pathname.match(exclusion);
        if (match !== null) {
          return next();
        }
      }
    }

    const pathname = parsedUrl.pathname;
    if (pathname.lastIndexOf('.') > pathname.lastIndexOf('/') &&
        options.disableDotRule !== true) {
      logger(
        'Not rewriting',
        req.method,
        req.url,
        'because the path includes a dot (.) character.'
      );
      return next();
    }

    rewriteTarget = options.index || '/index.html';
    logger('Rewriting', req.method, req.url, 'to', rewriteTarget);
    req.url = rewriteTarget;
    next();
  };
};

function evaluateRewriteRule(parsedUrl, match, rule, req) {
  if (typeof rule === 'string') {
    return rule;
  } else if (typeof rule !== 'function') {
    throw new Error('Rewrite rule can only be of type string or function.');
  }

  return rule({
    parsedUrl: parsedUrl,
    match: match,
    request: req
  });
}

function acceptsHtml(header, options) {
  options.htmlAcceptHeaders = options.htmlAcceptHeaders || ['text/html', '*/*'];
  for (let i = 0; i < options.htmlAcceptHeaders.length; i++) {
    if (header.indexOf(options.htmlAcceptHeaders[i]) !== -1) {
      return true;
    }
  }
  return false;
}

function getLogger(options) {
  if (options && options.logger) {
    return options.logger;
  } else if (options && options.verbose) {
    return console.log.bind(console);
  }
  return function(){};
}

function isArray(item) {
  if (!Array.isArray(item)) {
    throw new Error(`${arguments[0]} must be an array`);
  } else {
    return true;
  }
}
